File: Tundra/ZiaFile.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Public Module ZiaFile
    Public Enum LineType
        Comment
        KeyValue
        Blank
    End Enum
    Public Structure ZiaLine
        Public LineType As LineType
        Public Key As String
        Public Value As String
        Public Sub New(ByVal _LineType As LineType, Optional ByVal _Key As String = "", Optional ByVal _Value As String = "")
            LineType = _LineType
            Key = _Key
            Value = _Value
        End Sub
    End Structure

    Public Function ReadSectioned(ByVal Data As String) As Dictionary(Of String, Dictionary(Of String, String))
        On Error Resume Next
        Dim Sections As String() = Data.Split(New String() {"["}, StringSplitOptions.RemoveEmptyEntries)
        Dim TemporaryDictionary As New Dictionary(Of String, Dictionary(Of String, String))
        For Each Section As String In Sections
            TemporaryDictionary.Add(Section.Split(vbLf).GetValue(0).ToString.Remove(Section.Split(vbLf).GetValue(0).ToString.Length - 1, 1), Read(Section))
        Next
        Return TemporaryDictionary
    End Function

    Public Function Read(ByVal Data As String) As Dictionary(Of String, String)
        On Error Resume Next
        Dim Lines As String() = Data.Split(New String() {Environment.NewLine}, StringSplitOptions.RemoveEmptyEntries)
        Dim TemporaryDictionary As New Dictionary(Of String, String)
        For Each Line As String In Lines
            If Line.StartsWith("!") Then Continue For
            If Line.StartsWith("#") Then Continue For
            If Not Line.Contains("=") Then Continue For
            TemporaryDictionary.Add(FromCompatibleString(Line.Split("=").GetValue(0)), FromCompatibleString(Line.Split("=").GetValue(1)))
        Next
        Return TemporaryDictionary
    End Function
    Public Function Write(ByVal Source As List(Of ZiaLine), Optional ByVal MakeCompatible As Boolean = False) As String
        Dim RString As String = ""
        For Each Line As ZiaLine In Source
            If Line.LineType = LineType.Comment Then
                RString = RString.Insert(RString.Length, String.Format("#{0}", Line.Key))
            ElseIf Line.LineType = LineType.KeyValue Then
                If MakeCompatible = True Then
                    RString = RString.Insert(RString.Length, String.Format("{0}={1}", Line.Key, ToCompatibleString(Line.Value)))
                Else
                    RString = RString.Insert(RString.Length, String.Format("{0}={1}", Line.Key, Line.Value))
                End If
            Else

            End If
            RString = RString.Insert(RString.Length, Environment.NewLine)
        Next
        Return RString
    End Function

    Public Function SearchSectioned(ByVal Data As Dictionary(Of String, Dictionary(Of String, String)), ByVal LookupSection As String, ByVal LookupKey As String) As String
        On Error Resume Next
        For Each CurrentSection As KeyValuePair(Of String, Dictionary(Of String, String)) In Data
            If Not CurrentSection.Key = LookupSection Then Continue For
            Return Search(CurrentSection.Value, LookupKey)
        Next
    End Function

    Public Function Search(ByVal Data As Dictionary(Of String, String), ByVal Lookup As String) As String
        On Error Resume Next
        Return Data.Item(Lookup)
    End Function

    Public Function GetValue(ByVal Data As String, ByVal Lookup As String) As String
        On Error Resume Next
        Return ZiaFile.Search(ZiaFile.Read(Data), Lookup)
    End Function

    Public Function ToCompatibleString(ByVal Source As String) As String
        On Error Resume Next
        Source = Source.Replace("=", ":eq:")
        Source = Source.Replace("|", ":pi:")
        Source = Source.Replace(",", ":co:")
        Source = Source.Replace(Environment.NewLine, ":cr:")
        Source = Source.Replace(vbCrLf, ":cr:")
        Source = Source.Replace(vbLf, ":cr:")
        Source = Source.Replace(vbCr, ":cr:")
        Return Source
    End Function

    Public Function ToCompatibleColor(ByVal Source As Color) As String
        On Error Resume Next
        If Source = Color.Transparent Then
            Return "$0"
        ElseIf Source = Color.White Then
            Return "$1"
        ElseIf Source = Color.Black Then
            Return "$2"
        Else
            Return String.Format("{0},{1},{2},{3}", Source.A, Source.R, Source.G, Source.B)
        End If
    End Function

    Public Function ToCompatibleSingleList(ByVal Source() As Single) As String
        On Error Resume Next
        Dim CompatibleSingleList As String = ""
        For Each _Single As Single In Source
            CompatibleSingleList = CompatibleSingleList & _Single & "|"
        Next
        Return CompatibleSingleList
    End Function

    Public Function ToCompatibleStringList(ByVal Source As List(Of String)) As String
        On Error Resume Next
        Dim CompatibleSingleList As String = ""
        For Each _String As String In Source
            If _String = "" Then Continue For
            If _String = " " Then Continue For
            CompatibleSingleList = CompatibleSingleList & _String & "|"
        Next
        Return CompatibleSingleList
    End Function

    Public Function ToCompatibleColorList(ByVal Source() As Color) As String
        On Error Resume Next
        Dim CompatibleSingleList As String = ""
        For Each _Color As Color In Source
            CompatibleSingleList = CompatibleSingleList & ToCompatibleColor(_Color) & "|"
        Next
        Return CompatibleSingleList
    End Function

    Public Function ToCompatibleFont(ByVal Source As Font) As String
        On Error Resume Next
        Return String.Format("{0}|{1}|{2}", Source.Name, Source.Size, Source.Style)
    End Function

    Public Function FromCompatibleString(ByVal Source As String) As String
        On Error Resume Next
        Source = Source.Replace(":eq:", "=")
        Source = Source.Replace(":pi:", "|")
        Source = Source.Replace(":co:", ",")
        Source = Source.Replace(":cr:", Environment.NewLine)
        Return Source
    End Function

    Public Function FromCompatibleColorList(ByVal Source As String) As Color()
        On Error Resume Next
        Dim TempararyColors As New List(Of Color)
        Dim SplitColors As String() = Source.Split(New String() {"|"}, StringSplitOptions.RemoveEmptyEntries)
        For Each ThisColor As String In SplitColors
            If ThisColor = Nothing Then Return Nothing
            If Char.IsWhiteSpace(ThisColor.Chars(1)) = True Then Return Nothing
            TempararyColors.Add(FromCompatibleColor(ThisColor))
        Next
        Return TempararyColors.ToArray
    End Function

    Public Function FromCompatibleIntegerList(ByVal Source As String) As Integer()
        On Error Resume Next
        Dim TempararyIntegers As New List(Of Integer)
        Dim SplitIntegers As String() = Source.Split(New String() {"|"}, StringSplitOptions.RemoveEmptyEntries)
        For Each ThisInteger As String In SplitIntegers
            If ThisInteger = Nothing Then Return Nothing
            If Char.IsWhiteSpace(ThisInteger.Chars(1)) = True Then Return Nothing
            TempararyIntegers.Add(ThisInteger)
        Next
        Return TempararyIntegers.ToArray
    End Function

    Public Function FromCompatibleSingleList(ByVal Source As String) As Single()
        On Error Resume Next
        Dim TempararySingles As New List(Of Single)
        Dim SplitSingles As String() = Source.Split(New String() {"|"}, StringSplitOptions.RemoveEmptyEntries)
        If SplitSingles.Length = 0 Then Return Nothing
        For Each ThisSingle As String In SplitSingles
            If ThisSingle = Nothing Then Return Nothing
            If ThisSingle = "" Then Return Nothing
            If ThisSingle = " " Then Return Nothing
            TempararySingles.Add(ThisSingle)
        Next
        Return TempararySingles.ToArray
    End Function

    Public Function FromCompatibleColor(ByVal Source As String) As Color
        On Error Resume Next
        If Source.StartsWith("?") Then
            Return Color.FromArgb(Source.Split("/").GetValue(0).ToString.Remove(0, 1), FromCompatibleColor(Source.Split("/").GetValue(1)))
            Exit Function
        End If

        If Source.Split(",").Length = 4 Then
            Return Color.FromArgb(Source.Split(",").GetValue(0), Source.Split(",").GetValue(1), Source.Split(",").GetValue(2), Source.Split(",").GetValue(3))
        ElseIf Source.Split(",").Length = 3 Then
            Return Color.FromArgb(Source.Split(",").GetValue(0), Source.Split(",").GetValue(1), Source.Split(",").GetValue(2))
        ElseIf Source.StartsWith("#") Then
            Return ColorTranslator.FromHtml(Source)
        ElseIf Source.StartsWith("~") Then
            Dim B As String = Source.Remove(0, 1)
            Return Color.FromArgb(B, B, B)
        ElseIf Source = "$NULL" Then
            Return Color.Transparent
        ElseIf Source = "$0" Then
            Return Color.Transparent
        ElseIf Source = "$1" Then
            Return Color.White
        ElseIf Source = "$2" Then
            Return Color.Black
        Else
            Return Color.Transparent
        End If
    End Function
    Public Function FromCompatibleFont(ByVal Source As String) As Font
        Dim _Name As String = Source.Split("|").GetValue(0)
        Dim _Single As Single = CType((Source.Split("|").GetValue(1)), Single)
        Dim _FontStyle As FontStyle = Source.Split("|").GetValue(2)
        Return New Font(_Name, _Single, _FontStyle)
    End Function
    Public Function FromCompatibleFont2(ByVal Source As String) As Font
        Dim _Name As String = Source.Split("|").GetValue(0)
        Dim _Single As Single = CType((Source.Split("|").GetValue(1)), Single)
        Dim _FontStyle As FontStyle = CType([Enum].Parse(GetType(FontStyle), Source.Split("|").GetValue(2)), FontStyle)
        Return New Font(_Name, _Single, _FontStyle)
    End Function
    Public Function FromCompatiblePoint(ByVal Source As String) As Point
        Source = Source.Remove(0, 1)
        Source = Source.Remove(Source.Length - 1, 1)
        Return New Point(Source.Split(",").GetValue(0), Source.Split(",").GetValue(1))
    End Function

End Module

File: TundraLib/ZiaFile.cs
using System;
using System.Collections.Generic;
using System.Drawing;

namespace TundraLib
{
    public static class ZiaFile
    {
        public enum LineType
        {
            Comment,
            KeyValue,
            Blank
        }
        public struct ZiaLine
        {
            public LineType LineType;
            public string Key;
            public string Value;
            public ZiaLine(LineType _LineType, string _Key = "", string _Value = "")
            {
                LineType = _LineType;
                Key = _Key;
                Value = _Value;
            }
        }

        public static Dictionary<string, Dictionary<string, string>> ReadSectioned(string Data)
        {
            ;/* Cannot convert OnErrorResumeNextStatementSyntax, CONVERSION ERROR: Conversion for OnErrorResumeNextStatement not implemented, please report this issue in 'On Error Resume Next' at character 620
   at ICSharpCode.CodeConverter.CSharp.VisualBasicConverter.MethodBodyVisitor.DefaultVisit(SyntaxNode node)
   at Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor`1.VisitOnErrorResumeNextStatement(OnErrorResumeNextStatementSyntax node)
   at Microsoft.CodeAnalysis.VisualBasic.Syntax.OnErrorResumeNextStatementSyntax.Accept[TResult](VisualBasicSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor`1.Visit(SyntaxNode node)
   at ICSharpCode.CodeConverter.CSharp.CommentConvertingMethodBodyVisitor.ConvertWithTrivia(SyntaxNode node)
   at ICSharpCode.CodeConverter.CSharp.CommentConvertingMethodBodyVisitor.DefaultVisit(SyntaxNode node)

Input: 
        On Error Resume Next

 */
            string[] Sections = Data.Split(new string[] { "[" }, StringSplitOptions.RemoveEmptyEntries);
            Dictionary<string, Dictionary<string, string>> TemporaryDictionary = new Dictionary<string, Dictionary<string, string>>();
            foreach (string Section in Sections)
                TemporaryDictionary.Add(Section.Split('\n')[0].ToString().Remove(Section.Split('\n')[0].ToString().Length - 1, 1), Read(Section));
            return TemporaryDictionary;
        }

        public static Dictionary<string, string> Read(string Data)
        {
            ;/* Cannot convert OnErrorResumeNextStatementSyntax, CONVERSION ERROR: Conversion for OnErrorResumeNextStatement not implemented, please report this issue in 'On Error Resume Next' at character 1195
   at ICSharpCode.CodeConverter.CSharp.VisualBasicConverter.MethodBodyVisitor.DefaultVisit(SyntaxNode node)
   at Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor`1.VisitOnErrorResumeNextStatement(OnErrorResumeNextStatementSyntax node)
   at Microsoft.CodeAnalysis.VisualBasic.Syntax.OnErrorResumeNextStatementSyntax.Accept[TResult](VisualBasicSyntaxVisitor`1 visitor)
   at Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor`1.Visit(SyntaxNode node)
   at ICSharpCode.CodeConverter.CSharp.CommentConvertingMethodBodyVisitor.ConvertWithTrivia(SyntaxNode node)
   at ICSharpCode.CodeConverter.CSharp.CommentConvertingMethodBodyVisitor.DefaultVisit(SyntaxNode node)

Input: 
        On Error Resume Next

 */
            string[] Lines = Data.Split(new string[] { Environment.NewLine, "\r\n", "\r", "\n" }, StringSplitOptions.RemoveEmptyEntries);
            Dictionary<string, string> TemporaryDictionary = new Dictionary<string, string>();
            foreach (string Line in Lines)
            {
                if (Line.StartsWith("!"))
                    continue;
                if (Line.StartsWith("#"))
                    continue;
                if (!Line.Contains("="))
                    continue;
                TemporaryDictionary.Add(FromCompatibleString(Line.Split('=')[0]), FromCompatibleString(Line.Split('=')[1]));
            }
            return TemporaryDictionary;
        }

        public static string Write(List<ZiaLine> Source, bool MakeCompatible = false)
        {
            string RString = "";
            foreach (ZiaLine Line in Source)
            {
                if (Line.LineType == LineType.Comment)
                    RString = RString.Insert(RString.Length, string.Format("#{0}", Line.Key));
                else if (Line.LineType == LineType.KeyValue)
                {
                    if (MakeCompatible == true)
                        RString = RString.Insert(RString.Length, string.Format("{0}={1}", Line.Key, ToCompatibleString(Line.Value)));
                    else
                        RString = RString.Insert(RString.Length, string.Format("{0}={1}", Line.Key, Line.Value));
                }
                else
                {
                }
                RString = RString.Insert(RString.Length, Environment.NewLine);
            }
            return RString;
        }

        public static string SearchSectioned(Dictionary<string, Dictionary<string, string>> Data, string LookupSection, string LookupKey)
        {
            foreach (KeyValuePair<string, Dictionary<string, string>> CurrentSection in Data)
            {
                if (!(CurrentSection.Key == LookupSection))
                    continue;
                return Search(CurrentSection.Value, LookupKey);
            }

            return null;
        }

        public static string Search(Dictionary<string, string> Data, string Lookup)
        {
            return Data[Lookup];
        }

        public static string GetValue(string Data, string Lookup)
        {
            return Search(Read(Data), Lookup);
        }

        public static string ToCompatibleString(string Source)
        {
            Source = Source.Replace("=", ":eq:");
            Source = Source.Replace("|", ":pi:");
            Source = Source.Replace(",", ":co:");
            Source = Source.Replace(Environment.NewLine, ":cr:");
            Source = Source.Replace("\r\n", ":cr:");
            Source = Source.Replace("\n", ":cr:");
            Source = Source.Replace("\r", ":cr:");
            return Source;
        }

        public static string ToCompatibleColor(Color Source)
        {
            if (Source == Color.Transparent)
                return "$0";
            else if (Source == Color.White)
                return "$1";
            else if (Source == Color.Black)
                return "$2";
            else
                return string.Format("{0},{1},{2},{3}", Source.A, Source.R, Source.G, Source.B);
        }

        public static string ToCompatibleDoubleList(double[] Source)
        {
            string CompatibleDoubleList = "";
            foreach (double _Double in Source)
                CompatibleDoubleList = CompatibleDoubleList + _Double + '|';
            return CompatibleDoubleList;
        }

        public static string ToCompatibleStringList(List<string> Source)
        {
            string CompatibleDoubleList = "";
            foreach (string _String in Source)
            {
                if (_String == "")
                    continue;
                if (_String == " ")
                    continue;
                CompatibleDoubleList = CompatibleDoubleList + _String + '|';
            }
            return CompatibleDoubleList;
        }

        public static string ToCompatibleColorList(Color[] Source)
        {
            string CompatibleDoubleList = "";
            foreach (Color _Color in Source)
                CompatibleDoubleList = CompatibleDoubleList + ToCompatibleColor(_Color) + '|';
            return CompatibleDoubleList;
        }

        public static string ToCompatibleFont(Font Source)
        {
            return string.Format("{0}|{1}|{2}", Source.Name, Source.Size, Source.Style);
        }

        public static string FromCompatibleString(string Source)
        {
            Source = Source.Replace(":eq:", "=");
            Source = Source.Replace(":pi:", "|");
            Source = Source.Replace(":co:", ",");
            Source = Source.Replace(":cr:", Environment.NewLine);
            return Source;
        }

        public static Color[] FromCompatibleColorList(string Source)
        {
            List<Color> TempararyColors = new List<Color>();
            string[] SplitColors = Source.Split(new string[] { "|" }, StringSplitOptions.RemoveEmptyEntries);
            foreach (string ThisColor in SplitColors)
            {
                if (ThisColor == null)
                    return null;
                if (char.IsWhiteSpace(ThisColor[1]) == true)
                    return null;
                TempararyColors.Add(FromCompatibleColor(ThisColor));
            }
            return TempararyColors.ToArray();
        }

        public static int[] FromCompatibleIntegerList(string Source)
        {
            List<int> TempararyIntegers = new List<int>();
            string[] SplitIntegers = Source.Split(new string[] { "|" }, StringSplitOptions.RemoveEmptyEntries);
            foreach (string ThisInteger in SplitIntegers)
            {
                if (ThisInteger == null)
                    return null;
                if (char.IsWhiteSpace(ThisInteger[1]) == true)
                    return null;
                TempararyIntegers.Add(int.Parse(ThisInteger));
            }
            return TempararyIntegers.ToArray();
        }

        public static double[] FromCompatibleDoubleList(string Source)
        {
            List<double> TempararyDoubles = new List<double>();
            string[] SplitDoubles = Source.Split(new string[] { "|" }, StringSplitOptions.RemoveEmptyEntries);
            if (SplitDoubles.Length == 0)
                return null;
            foreach (string ThisDouble in SplitDoubles)
            {
                if (ThisDouble == null)
                    return null;
                if (ThisDouble == "")
                    return null;
                if (ThisDouble == " ")
                    return null;
                TempararyDoubles.Add(double.Parse(ThisDouble));
            }

            return TempararyDoubles.ToArray();
        }

        public static Color FromCompatibleColor(string Source)
        {
            if (Source.Split(',').Length == 4)
                return Color.FromArgb(int.Parse(Source.Split(',')[0]), int.Parse(Source.Split(',')[1]), int.Parse(Source.Split(',')[2]), int.Parse(Source.Split(',')[3]));
            else if (Source.Split(',').Length == 3)
                return Color.FromArgb(int.Parse(Source.Split(',')[0]), int.Parse(Source.Split(',')[1]), int.Parse(Source.Split(',')[2]));
            else if (Source.StartsWith("#"))
                return ColorTranslator.FromHtml(Source);
            else if (Source.StartsWith("~"))
            {
                int black = int.Parse(Source.Remove(0, 1));
                return Color.FromArgb(black, black, black);
            }
            else if (Source == "$NULL")
                return Color.Transparent;
            else if (Source == "$0")
                return Color.Transparent;
            else if (Source == "$1")
                return Color.White;
            else if (Source == "$2")
                return Color.Black;
            else
                return Color.Transparent;
        }

        public static Font FromCompatibleFont(string Source)
        {
            string _Name = Source.Split('|')[0];
            float _Size = (float.Parse(Source.Split('|')[1]));
            FontStyle _FontStyle = (FontStyle)Enum.Parse(typeof(FontStyle), Source.Split('|')[2]);
            return new Font(_Name, _Size, _FontStyle);
        }
        public static Point FromCompatiblePoint(string Source)
        {
            Source = Source.Remove(0, 1);
            Source = Source.Remove(Source.Length - 1, 1);
            return new Point(int.Parse(Source.Split(',')[0]), int.Parse(Source.Split(',')[1]));
        }
    }
}